File: android-compose/app/src/main/java/com/ohmz/tday/compose/core/data/server/VersionCompatibility.kt
package com.ohmz.tday.compose.core.data.server

import com.ohmz.tday.compose.BuildConfig
import com.ohmz.tday.compose.core.security.ProbeCompatibilityPayload
import com.ohmz.tday.compose.feature.release.compareVersions

data class VersionCompatibility(
    val appVersion: String,
    val updateRequired: Boolean,
)

sealed class VersionCheckResult {
    data object Compatible : VersionCheckResult()

    data class AppUpdateRequired(
        val requiredVersion: String,
    ) : VersionCheckResult()

    data class ServerUpdateRequired(
        val serverVersion: String,
    ) : VersionCheckResult()
}

fun checkVersionCompatibility(payload: ProbeCompatibilityPayload?): VersionCheckResult {
    if (payload == null) return VersionCheckResult.Compatible

    val localVersion = BuildConfig.VERSION_NAME
    val serverAppVersion = payload.appVersion
    val cmp = compareVersions(localVersion, serverAppVersion)

    if (!payload.updateRequired) return VersionCheckResult.Compatible

    return when {
        cmp < 0 -> VersionCheckResult.AppUpdateRequired(requiredVersion = serverAppVersion)
        cmp > 0 -> VersionCheckResult.ServerUpdateRequired(serverVersion = serverAppVersion)
        else -> VersionCheckResult.Compatible
    }
}

File: ios-swiftUI/Tday/Core/Model/VersionCompatibility.swift
import Foundation

struct VersionCompatibility: Equatable {
    let appVersion: String
    let updateRequired: Bool
}

enum VersionCheckResult: Equatable {
    case compatible
    case appUpdateRequired(requiredVersion: String)
    case serverUpdateRequired(serverVersion: String)
}

func checkVersionCompatibility(payload: ProbeCompatibilityPayload?) -> VersionCheckResult {
    guard let payload else { return .compatible }

    let localVersion = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0.0"
    let cmp = AppViewModel.compareVersions(localVersion, payload.appVersion)

    guard payload.updateRequired else { return .compatible }

    if cmp < 0 {
        return .appUpdateRequired(requiredVersion: payload.appVersion)
    } else if cmp > 0 {
        return .serverUpdateRequired(serverVersion: payload.appVersion)
    }
    return .compatible
}
